### DISCUSSION TITLE
Not able to load related fields with async sqlalchemy

### DISCUSSION THREAD

**Discussion by rbhanot4739** [sqlalchemy/sqlalchemy#10176] [Usage Questions]
Hello,

I am running Sqlalchemy 1.4.28 my with MySQL 8.0.19 with `asyncmy` as database driver and have the following model mapper classes defined

```py
class CustomBase(Timestamp):
    """
    Custom base class which will add
    created and updated fields automatically to the newly created table
    __repr__ method
    """

    def as_dict(self):
        return {k: v for k, v in vars(self).items() if not k.startswith("_sa")}


CustomModelBase: Any = declarative_base(cls=CustomBase)


class NewFaultMap(CustomModelBase):
    __tablename__ = "faultmap"

    id = Column("id", Integer, primary_key=True)
    check_id = Column(String(96), unique=True, nullable=False)

    # One to many relationship with Fault class
    faults = relationship("Fault", back_populates="faultmap", lazy="selectin")


class CM(CustomModelBase):
    __tablename__ = "cm"

    cm_id = Column("id", Integer(), primary_key=True)
    status = Column("status", String(32), nullable=False)
    created_by = Column("created_by", String(64), default="")
    updated_by = Column("updated_by", String(64), default="")


    # One to many relationship with Fault class
    faults = relationship(
        "Fault",
        back_populates="cm",
        lazy="selectin",
        cascade="all, delete-orphan",
    )


class Fault(CustomModelBase):
    __tablename__ = "fault"

    id = Column("id", Integer(), primary_key=True)
    cm_id = Column(
        Integer,
        ForeignKey("cm.id", ondelete="CASCADE"),
        index=True,
    )
    check_id = Column(
        String(96),
        ForeignKey("faultmap.check_id", onupdate="CASCADE"),
        nullable=False,
    )
    jira_ticket = Column("jira_ticket", String(64), default="")
    cm = relationship("CM", back_populates="faults", lazy="selectin")
    faultmap = relationship("NewFaultMap", back_populates="faults", lazy="selectin")
```

I am trying move our codebase to use the asyncio model and have gone through the docs [here](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-orm) and I am aware of some of the implications like [this](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession).

Maybe I am not interpreting the docs correctly but for some reason even though the loader strategy in my mapper class definition is specified as `selectin` on both sides of the `relationship`  but when I run 

`res = (await session.scalars(select(CM).where(CM.cm_id == 56487))).first()`

and then try to fetch the relationship attributes like  `res.faults[0].faultmap.faults`  it does not work and throws the typical `(sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place?`  error.

I have also tried to be explicit with  ` r = (await session.scalars(select(CM).options(selectinload(CM.faults).selectinload(Fault.faultmap)).where(*where_clause))).first()`  but to no success and backward relationship data does not gets loaded.

Is it possible to achieve this behaviour where the backward relationships are also loaded eagerly ?

**Comment by rbhanot4739** [comment]
Can someone please help me answering above ?

**Comment by zzzeek** [comment]
eager loaders like "selectinload" have limiting applied to avoid going into recursive loops, which also might be taking effect here.   I would not expect A->B->A with "selectinload" all the way through to load like that, since it would run the risk of a recursive loop occurring and this checking is obviously fairly conservative.   it might work if you put "joinedload" on the many-to-one side instead.

specifying the options directly should work without any recursion checking, however I see the path you are looking at is:

    res.faults[0].faultmap.faults

that is, it has three segments past the initial result

    selectinload(CM.faults).selectinload(Fault.faultmap)

has only two segments.  you'd need to list out the full path you want:

    selectinload(CM.faults).selectinload(Fault.faultmap).selectinload(WhateverThisIs.faults)


for many-to-one loads, using an identity map fetch using lazyload or immediateload might be a better approach which will often avoid running a second query.   with asyncio, lazyload is not an option however you might be able to use immediateload, however, I would want to wait for 2.0.20 for that as I [just fixed a bug for this use case](https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-a0f7cc7424b35990fc56876da2f33e94) as this same issue was just reported elsewhere.

I also want to note that people can get out of these issues right now for a particular place that a unwanted lazyload is occurring by simply awaiting on the lazy accessor, using the new [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs) API.
